### PR TITLE
Hash tables and queues for efficient lookup.

### PR DESCRIPTION
**Pull request by nirbhayc**

### CODE REVIEW

File: db/comdb2.c [hunk 1]
```diff
@@ -875,39 +874,33 @@ struct db *getdbbynum(int num)
...
-    int idx;
+    struct db *db;
+    int idx = -1;
 
-    /*should be changed to a hash table*/
-    for (idx = 0; idx < thedb->num_dbs; ++idx) {
-        if (strcasecmp(thedb->dbs[idx]->dbname, p_name) == 0)
-            return idx;
+    db = hash_find_readonly(thedb->db_hash, &p_name);
+    for (int i = 0; i < thedb->num_dbs && db; ++i) {
```
Comment (dorinhogea) [inline db/comdb2.c]: Please use db->dbs_idx with pointer validation instead of array scan

File: db/comdb2.c [hunk 2]
```diff
@@ -8805,38 +8820,16 @@ void delete_db(char *db_name)
         exit(1);
     }
 
+    /* Remove the table from hash. */
+    hash_del(thedb->db_hash, thedb->dbs[idx]);
+
```
Comment (dorinhogea) [inline db/comdb2.c]: What protects this hash against concurrent getdbbyname() lookups?

File: db/comdb2.c (hunk 1 above)
Comment (nirbhayc) [inline db/comdb2.c]: Even better. Thanks!

File: db/comdb2.c (hunk 2 above)
Comment (nirbhayc) [inline db/comdb2.c]: It's protected by delete_db()'s acquisition of wrlock on thedb_lock.

**Pull request merged**